src/ncp/rcp_host.cpp

```diff
@@ -595,9 +621,36 @@ void RcpHost::SetChannelMaxPowers(const std::vector<ChannelMaxPower> &aChannelMa
 }
 #endif // OTBR_ENABLE_POWER_CALIBRATION
 
-void RcpHost::DisableThreadAfterDetach(void *aContext)
+void RcpHost::ThreadDetachGracefully(const DetachGracefullyCallback &aCallback)
+{
+    mDetachGracefullyCallbacks.push_back(aCallback);
+
+    // Ignores the OT_ERROR_BUSY error if a detach has already been requested
+    OT_UNUSED_VARIABLE(otThreadDetachGracefully(mInstance, ThreadDetachGracefullyCallback, this));
+}
+
+void RcpHost::ThreadDetachGracefullyCallback(void *aContext)
+{
+    static_cast<RcpHost *>(aContext)->ThreadDetachGracefullyCallback();
+}
+
+void RcpHost::ThreadDetachGracefullyCallback(void)
 {
-    static_cast<RcpHost *>(aContext)->DisableThreadAfterDetach();
+    SafeInvokeAndClear(mScheduleMigrationReceiver, OT_ERROR_ABORT, "Aborted by leave/disable operation");
+
+    for (auto &callback : mDetachGracefullyCallbacks)
+    {
+        callback();
+    }
+    mDetachGracefullyCallbacks.clear();
+}
+
+void RcpHost::ErasePersistentInfo(bool aErase)
```

Comment:
Seems when `aErase` is false it's a no-op, which kind of conflicts with the method's name.

Reply:
What about naming it as `ConditionalErasePersistentInfo`? This is basically an inline function to write shorter code.

Reply:
SGTM name it as `ConditionalErasePersistentInfo`. In a longer term we may want to introduce a util for this pattern.